Full tutorial review day 01 early-task (#19)
epiverse-trace/tutorials-early: episodes/delays-reuse.Rmd

---

> title: 'Reuse delay data'

Reviewer: Wonder if this title could be more informative, e.g. 'Access epidemiological delay distributions'?

Author: added in latest commits

teaching: 10
exercises: 2
editor_options: 
  chunk_output_type: inline
---

:::::::::::::::::::::::::::::::::::::: questions 

- How to get easy access to delay distributions from a literature search database?

::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: objectives

- Get delays from a literature search database with `{epiparameter}`.
- Get distribution parameters and summary statistics of delay distributions.

::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: prereq

## Prerequisites

This episode requires you to be familiar with:

**Data science** : Basic programming with R.

> **Epidemic theory** : Epidemiological parameters, time periods.

Author: added in https://github.com/epiverse-trace/tutorials-early/pull/50/commits/ef5877d7d68fa357cd7406ed73d4e939e43f9834

:::::::::::::::::::::::::::::::::

## Introduction

The [natural history](../learners/reference.md#naturalhistory) of an infectious disease shows that its development has a regularity from stage to stage. The time periods from an infectious disease inform about the timing of transmission and interventions.

> ![Definition of key time periods. From [Xiang et al, 2021](https://www.sciencedirect.com/science/article/pii/S2468042721000038)](fig/time-periods.jpg)

Reviewer: Strictly speaking, I'd define this as 'presymptomatic' and 'symptomatic', with asymptomatic reserved for those who never develop symptoms (e.g. https://bmcinfectdis.biomedcentral.com/articles/10.1186/s12879-022-07440-0). But perhaps an edit for a later version – if needed, we could also make new schematics using some plots I've already made for various COVID presentations.

Author: Fair clarification, added as issue in #25

::::::::::::::::: callout

### Definitions

Look at the [glossary](../learners/reference.md) for the definitions of all the time periods of the figure above!

:::::::::::::::::::::::::

However, early in an epidemic, modelling efforts can be delayed by the lack of a centralised resource that summarises input parameters for the disease of interest ([Nash et al., 2023](https://mrc-ide.github.io/epireview/)). Projects like `{epiparameter}` and `{epireview}` are building online catalogues following literature synthesis protocols that can help parametrise models by easily accessing a comprenhensive library of previously estimated epidemiological parameters from past outbreaks.

<!-- Early models for COVID-19 used parameters from other coronaviruses. https://www.thelancet.com/article/S1473-3099(20)30144-4/fulltext -->

> To exemplify how to use `{epiparameter}` in your analysis pipeline, our goal in this episode will be to *choose* one specific set of epidemiological parameters from the literature, instead of *copying-and-pasting* them by hand, to plug them into an `{EpiNow2}` analysis workflow.

Author: appropriate edit included in https://github.com/epiverse-trace/tutorials-early/pull/50/commits/47a538508663b478a022467a249573de1e52b4ef

> To exemplify how to use `{epiparameter}` in your analysis pipeline, our goal in this episode will be to *choose* one specific set of epidemiological parameters from the literature, instead of *copying-and-pasting* them by hand, to plug them into an `{EpiNow2}` analysis workflow.

Author: included in https://github.com/epiverse-trace/tutorials-early/pull/50/commits/47a538508663b478a022467a249573de1e52b4ef

> To exemplify how to use `{epiparameter}` in your analysis pipeline, our goal in this episode will be to *choose* one specific set of epidemiological parameters from the literature, instead of *copying-and-pasting* them by hand, to plug them into an `{EpiNow2}` analysis workflow.

Author: I added the text edit suggested by Adam on the same topic

<!-- In this episode, we'll learn how to choose one specific set of epidemiological parameters from the literature and then get their **summary statistics** using `{epiparameter}`.  -->

Let's start loading the `{epiparameter}` package. We'll use the pipe `%>%` to connect some of their functions, some `{tibble}` and `{dplyr}` functions, so let's also call to the `{tidyverse}` package:

```{r,warning=FALSE,message=FALSE}

> library(epiparameter)

Author: this triggers an error because no EpiNow2 function is used here. We opted to add the namespace `EpiNow2::` before each function to overcome this. If they have it installed, this will run to learners locally.

library(tidyverse)
```

> ## The problem

Author: @CarmenTamayo what about `The goal of {epiparameter}`? The intention here is to introduce a problem to solve by epiparameter in this episode, which is to choose parameters from a pre-existing database instead of copy/paste.

> In `{EpiNow2}` we can specify a [generation time](../learners/reference.md#generationtime) as a probability `distribution` adding its `mean`, standard deviation (`sd`), and maximum value (`max`). To specify a `generation_time` that follows a _Gamma_ distribution with mean $\mu = 4$, standard deviation $\sigma = 2$, and a maximum value of 20, we write:

Reviewer: Think we should reiterate what EpiNow2 is and why someone would want to use it (especially as it's not loaded, so user can't run `dist_spec`). 

Also, are we keeping the `dist_spec` version for the alpha testing, or switching across to `Gamma()` functionality etc.?

```suggestion
If we want to estimate the transmissibility of an infection, it's common to use a package such as `{EpiEstim}` or `{EpiNow2}`. However, both require some epidemiological information as an input. For example, in `{EpiNow2}` we need to specify a [generation time](../learners/reference.md#generationtime) as a probability `distribution` adding its `mean`, standard deviation (`sd`), and maximum value (`max`). To specify a `generation_time` that follows a _Gamma_ distribution with mean $\mu = 4$, standard deviation $\sigma = 2$, and a maximum value of 20, we write:
```

Author: Agree to add that context.

We'll keep the `dist_spec` for this testing. dev team suggested to use the CRAN stable release. We'll wait for the next EpiNow2 release estimated for late April.

```r
generation_time <- dist_spec(
  mean = 4,
  sd = 2,
  max = 20,
  distribution = "gamma"
)
```

> Usually, we would *copy/paste* the **summary statistics** we found in a paper. Or, try to get them from the **distribution parameters** reported. One source of issue is that the report of different statistical distributions is not consistent across the literature. `{epiparameter}`’s objective is to provide information for a collection of distributions for a range of infectious diseases that is as accurate, unbiased and as comprehensive as possible.

Author: added in https://github.com/epiverse-trace/tutorials-early/pull/50/commits/b7c970ddc51104522cf924144f80698fb9ac5f99

<!-- https://epiverse-trace.github.io/epiparameter/articles/data_protocol.html -->

> Today, we'll *choose* the summary statistics from the library of epidemiological parameters provided by `{epiparameter}`.

Author: added in https://github.com/epiverse-trace/tutorials-early/pull/50/commits/a8c2407e4ed3d6a8f0449795c7a706b2e5dac874

> <!-- Instead of *manually* plug-in numeric values to `EpiNow2::dist_spec()` to specify the **summary statistics** of the delay distribution, we are going to *choose* them from the library of epidemiological parameters provided by `{epiparameter}`. -->

Author: text flanked by `<!-- text -->` is not shown. removed to clean text in https://github.com/epiverse-trace/tutorials-early/pull/50/commits/a8c2407e4ed3d6a8f0449795c7a706b2e5dac874

| Gamma | mean | shape | rate/scale |
|:--------------|:--------------|:--------------|:--------------|
| MERS-CoV | 14.13(13.9–14.7) | 6.31(4.88–8.52) | 0.43(0.33–0.60) |
| COVID-19 | 5.1(5.0–5.5) | 2.77(2.09–3.88) | 0.53(0.38–0.76) |

| Weibull | mean | shape | rate/scale |
|:--------------|:--------------|:--------------|:--------------|
| MERS-CoV | 14.2(13.3–15.2) | 3.07(2.64–3.63) | 16.1(15.0–17.1) |
| COVID-19 | 5.2(4.6–5.9) | 1.74(1.46–2.11) | 5.83(5.08–6.67) |

| Log normal | mean | mean-log | sd-log |
|:--------------|:--------------|:--------------|:--------------|
| MERS-CoV | 14.08(13.1–15.2) | 2.58(2.50–2.68) | 0.44(0.39–0.5) |
| COVID-19 | 5.2(4.2–6.5) | 1.45(1.31–1.61) | 0.63(0.54–0.74) |

Table: Serial interval estimates using Gamma, Weibull, and Log normal distributions. 95% confidence intervals for the shape and scale (logmean and sd for Log normal) parameters are shown in brackets ([Althobaity et al., 2022](https://www.sciencedirect.com/science/article/pii/S2468042722000537#tbl3)).

:::::::::::::::::::::::::

## Find a Generation time

> The generation time, jointly with the $R$, can inform about the speed of spread and its feasibility of control. Given a $R>1$, with a shorter generation time, cases can appear more quickly.

Author: added in https://github.com/epiverse-trace/tutorials-early/pull/50/commits/52f1efef385a8388add55fba3cb65712f379fded

![Video from the MRC Centre for Global Infectious Disease Analysis, Ep 76. Science In Context - Epi Parameter Review Group with Dr Anne Cori (27-07-2023) at <https://youtu.be/VvpYHhFDIjI?si=XiUyjmSV1gKNdrrL>](fig/reproduction-generation-time.png)

In calculating the effective reproduction number ($R_{t}$), the *generation time* distribution is often approximated by the [serial interval](../learners/reference.md#serialinterval) distribution.
This frequent approximation is because it is easier to observe and measure the onset of symptoms than the onset of infectiousness.

![A schematic of the relationship of different time periods of transmission between an infector and an infectee in a transmission pair. Exposure window is defined as the time interval having viral exposure, and transmission window is defined as the time interval for onward transmission with respect to the infection time ([Chung Lau et al., 2021](https://academic.oup.com/jid/article/224/10/1664/6356465)).](fig/serial-interval-observed.jpeg)

> However, using the *serial interval* as an approximation of the *generation time* is primarily valid for diseases in which infectiousness starts after symptom onset ([Chung Lau et al., 2021](https://academic.oup.com/jid/article/224/10/1664/6356465)). In cases where infectiousness starts before symptom onset, the serial intervals can have negative values, which is the case of a pre-symptomatic transmission ([Nishiura et al., 2020](https://www.ijidonline.com/article/S1201-9712(20)30119-3/fulltext#gr2)).

Author: added in https://github.com/epiverse-trace/tutorials-early/pull/50/commits/d98237fca4fde1a53227a83d21beb983b56b13f6

> However, using the *serial interval* as an approximation of the *generation time* is primarily valid for diseases in which infectiousness starts after symptom onset ([Chung Lau et al., 2021](https://academic.oup.com/jid/article/224/10/1664/6356465)). In cases where infectiousness starts before symptom onset, the serial intervals can have negative values, which is the case of a pre-symptomatic transmission ([Nishiura et al., 2020](https://www.ijidonline.com/article/S1201-9712(20)30119-3/fulltext#gr2)).

Author: the paragraph fits the topic but it is not developed further. I opted to block it temporarily to make it not visible. added in https://github.com/epiverse-trace/tutorials-early/pull/50/commits/b3a8bdc3465187bdd128c923621295ec6e61b046

> Additionally, even if the *generation time* and *serial interval* have the same mean, their variance usually differs, propagating bias to the $R_{t}$ estimation. $R_{t}$ estimates are sensitive not only to the mean generation time but also to the variance and form of the generation interval distribution [(Gostic et al., 2020)](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1008409).

Author: added in https://github.com/epiverse-trace/tutorials-early/pull/50/commits/d98237fca4fde1a53227a83d21beb983b56b13f6

::::::::::::::::::::::::::::::::: challenge

### Serial interval

Assume that COVID-19 and SARS have similar reproduction number values and that the serial interval approximates the generation time. 

Given the Serial interval of both infections in the figure below: 

- Which one would be harder to control? 
- Why do you conclude that?

![Serial interval of novel coronavirus (COVID-19) infections overlaid with a published distribution of SARS. ([Nishiura et al., 2020](https://www.ijidonline.com/article/S1201-9712(20)30119-3/fulltext))](fig/serial-interval-covid-sars.jpg)

::::::::::::::::: hint

The peak of each curve can inform you about the location of the mean of each distribution. The larger the mean, the larger the serial interval.

::::::::::::::::::::::

::::::::::::::::: solution

Which one would be harder to control?

- COVID-19

Why do you conclude that?

> - COVID-19 has the lowest mean serial interval. The approximate mean value for the serial interval of COVID-19 is around four days, and SARS is about seven days. Thus, COVID-19 will likely have newer generations in less time than SARS, assuming similar reproduction numbers.

Author: added in https://github.com/epiverse-trace/tutorials-early/pull/50/commits/d98237fca4fde1a53227a83d21beb983b56b13f6

::::::::::::::::::::::::::

:::::::::::::::::::::::::::::::::::::::::::

:::::::::::::::::::::: instructor

The objective of the assessment above is to assess the interpretation of a larger or shorter generation time.

::::::::::::::::::::::

> ## Extract epidemiological parameters

Author: added for consistency https://github.com/epiverse-trace/tutorials-early/pull/50/commits/d98237fca4fde1a53227a83d21beb983b56b13f6

> First, let's assume that the data set `example_confirmed` has COVID-19 observed cases. So, we need to find a reported generation time for COVID-19 or any other useful parameter for this aim. 

Author: good catch-up! mentioning the dataset name was a bug! added in https://github.com/epiverse-trace/tutorials-early/pull/50/commits/cfefcac0e98775f4947222ce1e3569a1fb7f88ef

> Let's start by looking at how many parameters we have in the epidemiological distributions database in `{epiparameter}` (`epidist_db`) for the `disease` named `covid`-19:

Author: added in https://github.com/epiverse-trace/tutorials-early/pull/50/commits/b3a8bdc3465187bdd128c923621295ec6e61b046

```{r}
epiparameter::epidist_db(
  disease = "covid"
)
```

From the `{epiparameter}` package, we can use the `epidist_db()` function to ask for any `disease` and also for a specific epidemiological distribution (`epi_dist`). 

Let's ask now how many parameters we have in the epidemiological distributions database (`epidist_db`) with the generation time using the string `generation`:

```{r}
epiparameter::epidist_db(
  epi_dist = "generation"
)
```

> Currently, in the library of epidemiological parameters, we have one `generation` time entry for Influenza. Considering the abovementioned considerations, we can look at the `serial` intervals for `COVID`-19. Run this locally!

Author: added in https://github.com/epiverse-trace/tutorials-early/pull/50/commits/b3a8bdc3465187bdd128c923621295ec6e61b046

```{r,eval=FALSE}
epiparameter::epidist_db(
  disease = "COVID",
  epi_dist = "serial"
)
```

With this query combination, we get more than one delay distribution. This output is an `<epidist>` class object.

::::::::::::::::: callout

> ### CASE-INSENSITIVE

Author: added in https://github.com/epiverse-trace/tutorials-early/pull/50/commits/b3a8bdc3465187bdd128c923621295ec6e61b046

`epidist_db` is [case-insensitive](https://dillionmegida.com/p/case-sensitivity-vs-case-insensitivity/#case-insensitivity). This means that you can use strings with letters in upper or lower case indistinctly.

:::::::::::::::::::::::::

To summarise an `<epidist>` object and get the column names from the underlying parameter database, we can add the `epiparameter::list_distributions()` function to the previous code using the pipe `%>%`:

```{r}
epiparameter::epidist_db(
  disease = "covid",
  epi_dist = "serial"
) %>%
  epiparameter::list_distributions()
```

In the `epiparameter::list_distributions()` output, we can also find different types of probability distributions (e.g., Log-normal, Weibull, Normal).

`{epiparameter}` uses the `base` R naming convention for distributions. This is why **Lognormal** is called `lnorm`.

::::::::::::::::: spoiler

> ### Why do we have a `<NA>` entry?

Author: ```suggestion
### Why do we have a <NA> entry?
```

> ### Why do we have a `<NA>` entry?

Author: yes, fixed now!

) %>%
  list_distributions()
```

We find two types of probability distributions for this query: _lognormal_ and _gamma_.

Now, search for delay distributions of your disease of interest! Ask:

- Do you recognise the papers?

- Should it consider any other paper?

How does `{epiparameter}` do the collection and review of peer-reviewed literature? We invite you to read the vignette on ["Data Collation and Synthesis Protocol"](https://epiverse-trace.github.io/epiparameter/articles/data_protocol.html)!

::::::::::::::::::::::::::

:::::::::::::::::::::::::::::::::::::::::::

## Select a single distribution

The `epiparameter::epidist_db()` function works as a filtering or subset function. Let's use the `author` argument to filter `Hiroshi Nishiura` parameters:

```{r}
epiparameter::epidist_db(
  disease = "covid",
  epi_dist = "serial",
  author = "Hiroshi"
) %>%
  epiparameter::list_distributions()
```

We still get more than one epidemiological parameter. We can set the `single_epidist` argument to `TRUE` to only one:

```{r}
epiparameter::epidist_db(
  disease = "covid",
  epi_dist = "serial",
  author = "Hiroshi",
  single_epidist = TRUE
)
```

::::::::::::::::: callout

> ### How does `single_epidist` works?

Reviewer: Formatting of this rendered a bit strangely for me on browser

![screenshot1](https://github.com/epiverse-trace/tutorials-early/assets/8329046/9fc1d1ab-132a-4631-8bc5-abea040a38ac)

Author: fixed

Looking at the help documentation for `?epiparameter::epidist_db()`:

- If multiple entries match the arguments supplied and `single_epidist = TRUE`, then the parameterised
`<epidist>` with the *largest sample size* will be returned.
- If multiple entries are equal after this sorting, the *first entry* will be returned.

What is a *parametrised* `<epidist>`? Look at `?is_parameterised`.

:::::::::::::::::::::::::

> Now, we have an epidemiological parameter we can reuse! We can replace the **summary statistics** numbers we plug into `EpiNow2::dist_spec()`.

Author: good catch-up again! missing bug. added in https://github.com/epiverse-trace/tutorials-early/pull/50/commits/b3a8bdc3465187bdd128c923621295ec6e61b046

```{r}
covid_serialint_parameters <- epiparameter::get_parameters(covid_serialint)

covid_serialint_parameters
```

This gets a vector of class `<numeric>` useful as input for any other package! 

:::::::::::::::::: callout

If we write the `[]` next to the last object create like in `covid_serialint_parameters[]`, within `[]` we can use the 
Tab key <kbd>↹</kbd> 
to use the [code completion feature](https://support.posit.co/hc/en-us/articles/205273297-Code-Completion-in-the-RStudio-IDE) and have a quick access to `covid_serialint_parameters["meanlog"]` and `covid_serialint_parameters["sdlog"]`. We invite you to try this out!

This only seems to work in code chunks and R console!

::::::::::::::::::

-->

You can use `plot()` to `<epidist>` objects to visualise:

- the *Probability Density Function (PDF)* and 
- the *Cumulative Distribution Function (CDF)*.

```{r}
# plot <epidist> object
plot(covid_serialint)
```

With the `day_range` argument, you can change the length or number of days in the `x` axis. Explore what this looks like:

```{r,eval=FALSE}
# plot <epidist> object
plot(covid_serialint, day_range = 0:20)
```

::::::::::::::::: discussion

> ### The distribution Zoo

Author: moved to the end from previous feedback

Explore this shinyapp called **The Distribution Zoo**!

Follow these steps to reproduce the form of the COVID serial interval distribution from `{epiparameter}` (`covid_serialint` object):

> 1. Access to <https://ben18785.shinyapps.io/distribution-zoo/> shiny app website,

Reviewer: ```suggestion
1. Access the <https://ben18785.shinyapps.io/distribution-zoo/> shiny app website,
```

2. Go to the left panel,
3. Keep the *Category of distribution*: `Continuous Univariate`,
4. Select a new *Type of distribution*: `Log-Normal`,
5. Move the **sliders**, i.e. the graphical control element that allows you to adjust a value by moving a handle along a horizontal track or bar to the `covid_serialint` parameters. 

Replicate these with the `distribution` object and all its list elements: `[[2]]`, `[[3]]`, and `[[4]]`. Explore how the shape of a distribution changes when its parameters change.

Share about:

- What other features of the website do you find helpful?

:::::::::::::::::::::::::

::::::::::::::::::::::::: instructor

In the context of user interfaces and graphical user interfaces (GUIs), like the [Distribution Zoo](https://ben18785.shinyapps.io/distribution-zoo/) shiny app, a **slider** is a graphical control element that allows users to adjust a value by moving a handle along a track or bar. Conceptually, it provides a way to select a numeric value within a specified range by visually sliding or dragging a pointer (the handle) along a continuous axis.

:::::::::::::::::::::::::

> ## Extract the summary statistics

Author: we rearrange it to get directly to the goal. if not yet, open to read more suggestions

```r
covid_serialint$
```

Use the `str()` to display the structure of the `<epidist>` R object.

::::::::::::::::::

:::::::::: solution

```{r,eval=TRUE}
# ebola serial interval
ebola_serial <-
  epiparameter::epidist_db(
    disease = "ebola",
    epi_dist = "serial",
    single_epidist = TRUE
  )

ebola_serial
```

```{r,eval=TRUE}
# get the sd
ebola_serial$summary_stats$sd

# get the sample_size
ebola_serial$metadata$sample_size
```

Try to visualise this distribution using `plot()`.

Also, explore all the other nested elements within the `<epidist>` object.

Share about: 

- What elements do you find useful for your analysis?
- What other elements would you like to see in this object? How?

::::::::::::::::::::

::::::::::::::::::::::::::::::::

:::::::::::::::::::::::::::::: instructor

An interesting element is the `method_assess` nested entry, which refers to the methods used by the study authors to assess for bias while estimating the serial interval distribution.

```{r}
covid_serialint$method_assess
```

We will explore these concepts following episodes!

::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::: challenge

> ### Ebola's severity parameter

Author: moved to end in previous added edits

Find a way to access the whole `{epiparameter}` database and find how that delay may be stored.

::::::::::::::::::::::

::::::::::::::::: solution

```{r,eval=TRUE}
# one way to get the list of all the available parameters
epidist_db(disease = "all") %>%
  list_distributions() %>%
  as_tibble() %>%
  distinct(epi_distribution)

ebola_severity <- epidist_db(
  disease = "ebola",
  epi_dist = "onset to discharge"
)

# point estimate
ebola_severity$summary_stats$mean
```

Check that for some `{epiparameter}` entries you will also have the *uncertainty* around the *point estimate* of each summary statistic:

```{r}
# 95% confidence intervals
ebola_severity$summary_stats$mean_ci
# limits of the confidence intervals
ebola_severity$summary_stats$mean_ci_limits
```

::::::::::::::::::::::::::

:::::::::::::::::::::::::::::::::::::::::::

Notice that with these pieces of information we can replace two out of three inputs of the `EpiNow2::dist_spec()` function:

```r
generation_time <- dist_spec(
  mean = covid_serialint$summary_stats$mean,
  sd = covid_serialint$summary_stats$sd,
  max = 20,
  distribution = "gamma"
)
```

> In the next episode we'll access to the `max` by using **distribution functions**!

Reviewer: I didn't understand this sentence I'm afraid.

Author: rephrased in some commits https://github.com/epiverse-trace/tutorials-early/pull/19/commits/722f1af9271f97c12a23549c96b6e5562a459430
